contrib/src/main/java/com/datatorrent/contrib/parser/StreamingJsonParser.java

```diff
@@ -0,0 +1,431 @@
+
+import com.google.common.collect.Lists;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.Context.OperatorContext;
+import com.datatorrent.api.Context.PortContext;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.annotation.OutputPortFieldAnnotation;
+import com.datatorrent.lib.parser.Parser;
+import com.datatorrent.lib.util.FieldInfo;
+import com.datatorrent.lib.util.FieldInfo.SupportType;
+import com.datatorrent.lib.util.KeyValPair;
+import com.datatorrent.lib.util.PojoUtils;
+
+/**
+ * Operator that parses a JSON string tuple and emits a POJO on the output port
+ * and tuples that could not be parsed on error port.Upstream operator needs to
+ * ensure that a full JSON record is emitted.<br>
+ * <b>Properties</b><br>
+ * <b>pojoClass</b>:POJO class <br>
+ * <b>(optional)fieldMappingString</b>String of format
+ * fieldNameInJson:fieldNameInPOJO:DataType<br>
+ * <b>Ports</b> <br>
+ * <b>in</b>:input tuple as a String. Each tuple represents a json string<br>
+ * <b>out</b>:tuples that are validated as per the user defined POJO are emitted
+ * as POJO on this port<br>
+ * <b>err</b>:tuples that could not be parsed are emitted on this port as
+ * KeyValPair<String,String><br>
+ * Key being the tuple and Val being the reason
+ * 
+ * @displayName SimpleStreamingJsonParser
+ * @category Parsers
+ * @tags json pojo parser streaming
+ */
+@InterfaceStability.Evolving
+public class StreamingJsonParser extends Parser<byte[], KeyValPair<String, String>>
+{
+  private transient JSONParser jsonParser;
+  private transient String fieldMappingString;
+  private transient List<FieldInfo> fieldInfos;
+  private transient List<ActiveFieldInfo> columnFieldSetters;
+  protected JsonKeyFinder finder;
+  private static final String FIELD_SEPARATOR = ":";
+  private static final String RECORD_SEPARATOR = ",";
+  private transient ArrayList<String> columnFields;
+  private transient Class<?> pojoClass;
+
+  /**
+   * @return POJO class
+   */
+  private Class<?> getPojoClass()
+  {
+    return pojoClass;
+  }
+
+  /**
+   * Sets the POJO class
+   */
+  public void setPojoClass(Class<?> pojoClass)
+  {
+    this.pojoClass = pojoClass;
+  }
+
+  /**
+   * Returns a string representing mapping from generic record to POJO fields
+   */
+  public String getFieldMappingString()
+  {
+    return fieldMappingString;
+  }
+
+  /**
+   * Comma separated list mapping a field in JSON schema to POJO field eg :
+   * fieldNameInPOJO:fieldNameInJSON:DataType
+   */
+  public void setFieldMappingString(String pojoFieldsToJsonMapping)
+  {
+    this.fieldMappingString = pojoFieldsToJsonMapping;
+  }
+
+  public StreamingJsonParser()
+  {
+
+  }
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    jsonParser = new JSONParser();
+    finder = new JsonKeyFinder();
+    columnFields = new ArrayList<String>();
+    columnFieldSetters = Lists.newArrayList();
+
+    ListIterator<FieldInfo> itr = fieldInfos.listIterator();
```

Comment:
Test seems to be working fine in my local mode including the application
test.
On May 27, 2016 12:00 AM, "shubham-pathak22" <notifications@github.com>
wrote:

> In
> contrib/src/main/java/com/datatorrent/contrib/parser/StreamingJsonParser.java
> <https://github.com/apache/incubator-apex-malhar/pull/288#discussion_r64864410>
> :
>
> > +  }
> > +
> > +  public StreamingJsonParser()
> > +  {
> > +
> > +  }
> > +
> > +  @Override
> > +  public void setup(OperatorContext context)
> > +  {
> > +    jsonParser = new JSONParser();
> > +    finder = new JsonKeyFinder();
> > +    columnFields = new ArrayList<String>();
> > +    columnFieldSetters = Lists.newArrayList();
> > +
> > +    ListIterator<FieldInfo> itr = fieldInfos.listIterator();
>
> Possible NPE. FieldInfos is getting initialized in setup of output port
> which gets called after this. Move this to setup method of output port
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-apex-malhar/pull/288/files/cf4af853e8dbf4c8876641d16230f463c35919e4#r64864410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHT673goi_W7dQf3_OSaD9rBGNpEgR5Jks5qFpZugaJpZM4ImDiC>
> .
>